File: backend/app/services/notifications.py
"""Notification dispatch.

In-app notifications are written to the database today. Email, SMS and push are
deliberately modelled as additional *channels* over the same records rather than
a separate mechanism: adding a delivery worker later means implementing
:class:`NotificationChannelSender` and registering it, with no change to the
call sites that raise the events.
"""

from __future__ import annotations

import logging
import uuid
from dataclasses import dataclass
from typing import Protocol, runtime_checkable

from sqlalchemy.orm import Session

from app.core.enums import NotificationChannel, UserRole
from app.models.system import Notification
from app.models.user import User
from app.repositories.system import NotificationRepository
from app.repositories.user import UserRepository

logger = logging.getLogger("roadwatch.notifications")


@dataclass(slots=True)
class NotificationPayload:
    event: str
    title: str
    body: str
    link: str | None = None
    complaint_id: uuid.UUID | None = None


@runtime_checkable
class NotificationChannelSender(Protocol):
    """Delivers a notification over one channel."""

    channel: NotificationChannel

    def send(self, user: User, payload: NotificationPayload) -> None:
        ...


class InAppSender:
    """Persists the notification for the in-app inbox."""

    channel = NotificationChannel.IN_APP

    def __init__(self, db: Session) -> None:
        self.repository = NotificationRepository(db)

    def send(self, user: User, payload: NotificationPayload) -> None:
        self.repository.add(
            Notification(
                user_id=user.id,
                title=payload.title,
                body=payload.body,
                event=payload.event,
                channel=self.channel,
                complaint_id=payload.complaint_id,
                link=payload.link,
            )
        )


class LoggingSender:
    """Stand-in for a real email/SMS gateway.

    Writes what *would* have been sent to the application log, so the
    integration point is exercised in development without a paid provider.
    """

    def __init__(self, channel: NotificationChannel) -> None:
        self.channel = channel

    def send(self, user: User, payload: NotificationPayload) -> None:
        logger.info(
            "[%s] to %s: %s - %s", self.channel.value, user.email, payload.title, payload.body
        )


class NotificationService:
    def __init__(self, db: Session, senders: list[NotificationChannelSender] | None = None) -> None:
        self.db = db
        self.users = UserRepository(db)
        self.senders = senders or [InAppSender(db)]

    def notify(self, user: User | None, payload: NotificationPayload) -> None:
        if user is None or not user.is_active:
            return
        for sender in self.senders:
            try:
                sender.send(user, payload)
            except Exception:  # noqa: BLE001 - delivery must never break the request
                logger.exception("Notification delivery failed on channel %s", sender.channel)

    def notify_user_id(self, user_id: uuid.UUID | None, payload: NotificationPayload) -> None:
        if user_id is None:
            return
        self.notify(self.users.get(user_id), payload)

    def notify_role(self, role: UserRole, payload: NotificationPayload) -> None:
        for user in self.users.list_by_role(role):
            self.notify(user, payload)

    def notify_team(self, team_id: uuid.UUID, payload: NotificationPayload) -> None:
        for user in self.users.list_by_team(team_id):
            self.notify(user, payload)

    # -- domain events --------------------------------------------------

    def complaint_submitted(self, complaint) -> None:
        link = f"/admin/reports/{complaint.id}"
        self.notify_role(
            UserRole.ADMIN,
            NotificationPayload(
                event="complaint.created",
                title=f"New {complaint.priority_level.value.lower()} priority report",
                body=(
                    f"{complaint.complaint_number}: "
                    f"{_label(complaint.damage_type)} reported"
                    f"{f' on {complaint.road_name}' if complaint.road_name else ''} "
                    f"with a priority score of {complaint.priority_score:g}."
                ),
                link=link,
                complaint_id=complaint.id,
            ),
        )

    def complaint_assigned(self, complaint, team_id: uuid.UUID, team_name: str) -> None:
        self.notify_team(
            team_id,
            NotificationPayload(
                event="assignment.created",
                title=f"New {complaint.priority_level.value.lower()} priority repair assigned",
                body=(
                    f"{complaint.complaint_number} ({_label(complaint.damage_type)}) "
                    f"has been assigned to your team."
                ),
                link=f"/team/tasks/{complaint.id}",
                complaint_id=complaint.id,
            ),
        )
        self.notify_user_id(
            complaint.reporter_id,
            NotificationPayload(
                event="complaint.assigned",
                title="Your report has been assigned",
                body=f"Your report {complaint.complaint_number} has been assigned to {team_name}.",
                link=f"/reports/{complaint.id}",
                complaint_id=complaint.id,
            ),
        )

    def complaint_status_changed(self, complaint, previous: str) -> None:
        self.notify_user_id(
            complaint.reporter_id,
            NotificationPayload(
                event="complaint.status_changed",
                title=f"Report {complaint.complaint_number} is now {_readable(complaint.status.value)}",
                body=(
                    f"Your reported {_label(complaint.damage_type)} moved from "
                    f"{_readable(previous)} to {_readable(complaint.status.value)}."
                ),
                link=f"/reports/{complaint.id}",
                complaint_id=complaint.id,
            ),
        )

    def complaint_resolved(self, complaint) -> None:
        self.notify_user_id(
            complaint.reporter_id,
            NotificationPayload(
                event="complaint.resolved",
                title="Your report has been resolved",
                body=(
                    f"The {_label(complaint.damage_type)} you reported "
                    f"({complaint.complaint_number}) has been repaired and verified. Thank you."
                ),
                link=f"/reports/{complaint.id}",
                complaint_id=complaint.id,
            ),
        )

    def repair_completed(self, complaint, team_name: str) -> None:
        self.notify_role(
            UserRole.ADMIN,
            NotificationPayload(
                event="repair.completed",
                title="Repair awaiting verification",
                body=(
                    f"{team_name} marked {complaint.complaint_number} complete and uploaded evidence. "
                    "Please verify."
                ),
                link=f"/admin/reports/{complaint.id}",
                complaint_id=complaint.id,
            ),
        )


def _label(damage_type) -> str:
    return str(damage_type.value if hasattr(damage_type, "value") else damage_type).replace(
        "_", " "
    ).lower()


def _readable(status: str) -> str:
    return status.replace("_", " ").lower()

File: supabase/functions/api/services/notifications.ts
// Ported from backend/app/services/notifications.py. Only the in-app
// channel is implemented (matches Python's default `senders=[InAppSender]`
// - email/SMS/push are modelled there as a Protocol with no real sender
// registered yet). Always uses the service client: notifications has no
// client-facing insert policy (RPC/trigger-only, see the RLS migration), and
// listing "every admin" or "every member of a team" must see across the
// caller's own visibility.

import type { SupabaseClient } from "@supabase/supabase-js";
import type { ComplaintRow } from "../repositories/complaint.ts";

interface NotificationPayload {
  event: string;
  title: string;
  body: string;
  link?: string | null;
  complaintId?: string | null;
}

function label(damageType: string): string {
  return damageType.replace(/_/g, " ").toLowerCase();
}

function readable(status: string): string {
  return status.replace(/_/g, " ").toLowerCase();
}

function trimNumber(value: number): string {
  return Number(value.toFixed(6)).toString();
}

async function insertNotification(client: SupabaseClient, userId: string, payload: NotificationPayload): Promise<void> {
  const { error } = await client.from("notifications").insert({
    user_id: userId,
    title: payload.title,
    body: payload.body,
    event: payload.event,
    channel: "IN_APP",
    complaint_id: payload.complaintId ?? null,
    link: payload.link ?? null,
  });
  if (error) {
    // Delivery must never break the request that triggered it.
    console.error("notification insert failed", error);
  }
}

async function notifyRole(client: SupabaseClient, role: string, payload: NotificationPayload): Promise<void> {
  const { data, error } = await client.from("profiles").select("id").eq("role", role).eq("is_active", true);
  if (error) throw error;
  await Promise.all((data ?? []).map((row) => insertNotification(client, row.id, payload)));
}

async function notifyTeam(client: SupabaseClient, teamId: string, payload: NotificationPayload): Promise<void> {
  const { data, error } = await client.from("profiles").select("id").eq("team_id", teamId).eq("is_active", true);
  if (error) throw error;
  await Promise.all((data ?? []).map((row) => insertNotification(client, row.id, payload)));
}

async function notifyUserId(client: SupabaseClient, userId: string | null | undefined, payload: NotificationPayload): Promise<void> {
  if (!userId) return;
  const { data, error } = await client.from("profiles").select("id, is_active").eq("id", userId).maybeSingle();
  if (error) throw error;
  if (!data || !data.is_active) return;
  await insertNotification(client, userId, payload);
}

export async function complaintSubmitted(client: SupabaseClient, complaint: ComplaintRow): Promise<void> {
  await notifyRole(client, "ADMIN", {
    event: "complaint.created",
    title: `New ${String(complaint.priority_level).toLowerCase()} priority report`,
    body:
      `${complaint.complaint_number}: ${label(String(complaint.damage_type))} reported` +
      (complaint.road_name ? ` on ${complaint.road_name}` : "") +
      ` with a priority score of ${trimNumber(Number(complaint.priority_score))}.`,
    link: `/admin/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function complaintAssigned(
  client: SupabaseClient,
  complaint: ComplaintRow,
  assignmentId: string,
  teamId: string,
  teamName: string,
): Promise<void> {
  await notifyTeam(client, teamId, {
    event: "assignment.created",
    title: `New ${String(complaint.priority_level).toLowerCase()} priority repair assigned`,
    body: `${complaint.complaint_number} (${label(String(complaint.damage_type))}) has been assigned to your team.`,
    // The crew's job page is keyed by the assignment's own id, not the
    // complaint's - the two are different rows, and a link built from the
    // wrong one 404s the moment it's clicked.
    link: `/team/tasks/${assignmentId}`,
    complaintId: complaint.id,
  });
  await notifyUserId(client, complaint.reporter_id, {
    event: "complaint.assigned",
    title: "Your report has been assigned",
    body: `Your report ${complaint.complaint_number} has been assigned to ${teamName}.`,
    link: `/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function complaintStatusChanged(client: SupabaseClient, complaint: ComplaintRow, previous: string): Promise<void> {
  await notifyUserId(client, complaint.reporter_id, {
    event: "complaint.status_changed",
    title: `Report ${complaint.complaint_number} is now ${readable(String(complaint.status))}`,
    body:
      `Your reported ${label(String(complaint.damage_type))} moved from ${readable(previous)} to ` +
      `${readable(String(complaint.status))}.`,
    link: `/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function complaintResolved(client: SupabaseClient, complaint: ComplaintRow): Promise<void> {
  await notifyUserId(client, complaint.reporter_id, {
    event: "complaint.resolved",
    title: "Your report has been resolved",
    body:
      `The ${label(String(complaint.damage_type))} you reported (${complaint.complaint_number}) has been ` +
      "repaired and verified. Thank you.",
    link: `/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function complaintDeleted(client: SupabaseClient, complaint: ComplaintRow, reason: string): Promise<void> {
  await notifyUserId(client, complaint.reporter_id, {
    event: "complaint.deleted",
    title: "Your report was removed",
    body: `${complaint.complaint_number} was removed by an administrator: ${reason}`,
    complaintId: complaint.id,
  });
}

export async function repairCompleted(client: SupabaseClient, complaint: ComplaintRow, teamName: string): Promise<void> {
  await notifyRole(client, "ADMIN", {
    event: "repair.completed",
    title: "Repair awaiting verification",
    body: `${teamName} marked ${complaint.complaint_number} complete and uploaded evidence. Please verify.`,
    link: `/admin/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function reworkRequested(
  client: SupabaseClient,
  complaint: ComplaintRow,
  assignmentId: string,
  teamId: string,
  reason: string,
): Promise<void> {
  await notifyTeam(client, teamId, {
    event: "repair.rework_requested",
    title: "Rework requested",
    body: `The works department sent ${complaint.complaint_number} back: ${reason}`,
    link: `/team/tasks/${assignmentId}`,
    complaintId: complaint.id,
  });
}

export async function repairFlagged(
  client: SupabaseClient,
  complaint: ComplaintRow,
  teamName: string,
  reason: string,
): Promise<void> {
  await notifyRole(client, "ADMIN", {
    event: "repair.flagged",
    title: "Crew flagged a problem with an assignment",
    body: `${teamName} could not proceed with ${complaint.complaint_number}: ${reason}. It is back in the queue, unassigned.`,
    link: `/admin/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}

export async function assignmentEscalated(
  client: SupabaseClient,
  complaint: ComplaintRow,
  reason: string,
): Promise<void> {
  await notifyRole(client, "ADMIN", {
    event: "repair.escalated",
    title: "🚨 Crew escalated a report as an emergency",
    body: `${complaint.complaint_number}: ${reason}`,
    link: `/admin/reports/${complaint.id}`,
    complaintId: complaint.id,
  });
}
